.gitlab-ci.yml

```diff
@@ -15,137 +16,189 @@ variables:
-    fi
-  - echo ${BRANCH}
-  - echo ${REPO}
   - git clone --depth 1 --branch ${BRANCH} ${REPO}
 
-.start-canvas:                     &start-canvas
-  - canvas --tmp --dev -lruntime::contracts=debug > /tmp/canvas.log 2>&1 &
+.start-substrate-contracts-node:                     &start-substrate-contracts-node
+  - substrate-contracts-node --tmp --dev -lruntime::contracts=debug > /tmp/substrate-contracts-node.log 2>&1 &
+
+.start-substrate-contracts-node-rand-extension:      &start-substrate-contracts-node-rand-extension
+  - substrate-contracts-node-rand-extension --tmp --dev -lruntime::contracts=debug > /tmp/substrate-contracts-node-rand-extension.log 2>&1 &
+
+.shutdown-substrate-contracts-node:                  &shutdown-substrate-contracts-node
+  - pkill -f "substrate-contracts-node --tmp --dev"
 
-.start-canvas-rand-extension:      &start-canvas-rand-extension
-  - canvas-rand-extension --tmp --dev -lruntime::contracts=debug > /tmp/canvas-rand-extension.log 2>&1 &
+.shutdown-substrate-contracts-node-rand-extension:   &shutdown-substrate-contracts-node-rand-extension
+  - pkill -f "substrate-contracts-node-rand-extension --tmp --dev"
 
-.build-contracts:                  &build-contracts
+
+# needed vars have to be "exported" on a earlier stage
+parent-vars:
+  stage:                                             declare-vars
+  <<:                                                *kubernetes-env
+  script:
+    - BRANCH="master"
+    - REPO="https://github.com/paritytech/ink.git"
+    - REPO_NAME="ink"
+    - echo ${TRGR_REF}
+    - if [ -n "$TRGR_REF" ] && [ "$TRGR_REF" != "master" ]; then
+        PR_JSON=`curl -s https://api.github.com/repos/paritytech/ink/pulls/${TRGR_REF}`;
+        BRANCH=`echo "${PR_JSON}" | jq -r .head.ref`;
+        REPO=`echo "${PR_JSON}" | jq -r .head.repo.git_url`;
+        REPO_NAME=`echo "${PR_JSON}" | jq -r .head.repo.name`;
+      fi
+    - echo "BRANCH=${BRANCH}" | tee -a parent-vars.env
+    - echo "REPO=${REPO}" | tee -a parent-vars.env
+    - echo "REPO_NAME=${REPO_NAME}" | tee -a parent-vars.env
+    - echo "CSV_NAME=${REPO_NAME}_${BRANCH}" | tee -a parent-vars.env
+    - echo "REDIS_RESULTS_KEY=${CI_PROJECT_NAME}::results::{$REPO_NAME}::${BRANCH}" | tee -a parent-vars.env
+    - echo "REDIS_RESULTS_KEY_MASTER=${CI_PROJECT_NAME}::results::{$REPO_NAME}::master" | tee -a parent-vars.env
+  artifacts:
+    reports:
+      dotenv: parent-vars.env
+
+
+.build-contracts:                                    &build-contracts
+  # delete old list items if the key has existed previously
+  - redis-cli -u $GITLAB_REDIS_URI del $REDIS_RESULTS_KEY
+  - echo 'Data will be written to $REDIS_RESULTS_KEY'
   - for example in ${INK_EXAMPLES_PATH}/*/; do
-      echo "cargo contract build --manifest-path=${example}/Cargo.toml" >> /tmp/cmds;
+      echo "./.scripts/build-contract.sh ${example} | redis-cli -u ${GITLAB_REDIS_URI} -x rpush ${REDIS_RESULTS_KEY}" >> /tmp/cmds;
```

Comment:
Just so I understand, will the following be what this would look like after a run?

```
> LRANGE project-name::results::ink::hc-test-branch 0 -1
1) "Contract A, 6, 3"
2) "Contract B, 4, 2"
...
```

Reply:
Yes, that's what it would be